### ISSUE TITLE
Get access to webpack options when the custom webpack returns a simple object

### ISSUE THREAD

**Issue by jeetparikh** [meltedspark/angular-builders#342]
It would be nice to have access to the options in the custom webpack config when it is returning an object and not a function.

For eg, when you run
`ng serve --open --configuration local --port 9000`
It would good to know inside the custom webpack config what configuration was passed.

The use case of this would be the use of `HtmlWebpackPlugin` which passes this configuration further to index.html which uses to switch few imports for product, development etc.

**Describe alternatives you've considered**
Currently I don't see any way around getting this problem. When using webpack without cli the environment variables are available in config files.

**Additional context**
angular: 8.0.0
angular-cli: 8.0.2
custom-webpack: 8.0.2

**Comment by jeetparikh** [comment]
@meltedspark if you could please confirm I am not missing anything and this is indeed a feature request. Thanks.

**Comment by meltedspark** [comment]
I'm not sure this is a valid use case (specifically the one you mentioned).  
`html-webpack-plugin` hasn’t been used by the CLI for `index.html` generation since version 1.7.  
Since version 8 they moved `index.html` generation completely out of the Webpack build (see [migration guide](https://github.com/meltedspark/angular-builders/blob/master/MIGRATION.MD)).  
In any case I don't see any added value to have it passed in any way to custom webpack config when it is returning an object. You already have this functionality in function-like webpack config.
If you're strongly against returning a function from your webpack config you can set an environment variable before you run the build (like [here](https://github.com/meltedspark/electron-angular-native/blob/master/package.json#L42)) and use it in your webpack config object (like [here](https://github.com/meltedspark/electron-angular-native/blob/master/native-webpack.config.js#L3)).

**Comment by jeetparikh** [comment]
thank you for your response @meltedspark ...
Well, I see it's deprecated. We are actually moving back to angular cli from webpack after an initial `ng eject` mainly because of this awesome builder that still allows to provide some custom config. Thanks for the great work.
When you say it is not valid use-case I am just trying to understand why. Here is my sample snippet from custom config
```
module.exports = {
    module: {
        rules: [
           // sass resouces loader
        ]
    },
    plugins: [
      // couple of other plugins
        new HtmlWebpackPlugin({
            template: './src/index.html',
            title: METADATA.title,
           env: env, // this passes in the environment to index.html which was possible through webpack
            version: pkg.version
        }),
```

and then using this dynamic data injected into `index.html`, it switches some imports for `*.min.js`, some `tracking`, some `logging` etc.. Why is this a not a valid scenario? just trying to think out load and understand why...
Do you think the the usage you have shown [here](https://github.com/meltedspark/electron-angular-native/blob/master/package.json#L42) which uses `cross-env` is the way to go for the use case till the `transform function` u mentioned in deprecation guide is released by angular cli?

Thanks @meltedspark

**Comment by meltedspark** [comment]
As I mentioned, the `index.html` generation is now not part of the Webpack build in `ng build`. They run it after the Webpack build is finished so your usage of `HtmlWebpackPlugin` won't affect final `dist/index.html`. It will only work for `ng serve` (which will change as well after the `transform` is ready).
So, to sum it up:
1. For `ng serve` use custom webpack config that returns a function. It's not a big deal, just wrap your object with a function. I advise against the `cross-env` solution, it is hacky.
2. For `ng build` for now you don't have another option but modifying `dist/index.html` after the build was finished. You'll have to write this code anyways, because once `transform` function is ready you'll have to alter `index.html` with no help of any Webpack plugin.

**Comment by jeetparikh** [comment]
Thanks for your response @meltedspark . appreciate it.

> For ng serve use custom webpack config that returns a function. It's not a big deal, just wrap your object with a function. I advise against the cross-env solution, it is hacky
If I use a function rather than an object, doesn't it mean, it will completely override the config provided internally used by angulars devkit.

> For ng build for now you don't have another option but modifying dist/index.html after the build was finished. You'll have to write this code anyways, because once transform function is ready you'll have to alter index.html with no help of any Webpack plugin.
I will consider doing this thanks. We also have some other custom requirements which I need to think about.

Thanks for your help @meltedspark

**Comment by meltedspark** [comment]
>If I use a function rather than an object, doesn't it mean, it will completely override the config provided internally used by angulars devkit.

In a way. You receive the original webpack config as input and you return a new one as an output. So instead of specifying the delta you modify the config you received as input. Like this:

```js
cfg.plugins.push({YOUR PLUGIN HERE});
cfg.module.rules.push({YOUR LOADER HERE});
```

>I will consider doing this thanks. We also have some other custom requirements which I need to think about.

Sure, if you want to discuss it you're welcome to [Slack](https://join.slack.com/t/angular-builders/shared_invite/enQtNTk5MDE1Nzg2ODM3LTkzMzI5ODQwYzRkOGZhZDJkNWI3OGJjNTM4YzNlY2MxZGU0ZjQ3NzVjZGEzNzYzNDFlOWM2YmQwZGQ3NzlkOWY).

**Comment by jeetparikh** [comment]
Thanks @meltedspark .. that was very helpful :)
